File: WinPaletter/GUI/Elements/Checkers/Toggle.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D

Namespace UI.WP

    <Description("Themed toggle for WinPaletter UI")> <DefaultEvent("CheckedChanged")> Public Class Toggle : Inherits UserControl

        Dim CheckC As New Rectangle(4, 4, 11, 11)
        Dim MouseState As Integer = 0
        Dim WasMoving As Boolean = False

        Sub New()
            DoubleBuffered = True
            Size = New Size(40, 20)
            Text = ""
        End Sub

        Public Property DarkLight_Toggler As Boolean = False

        ReadOnly DarkLight_TogglerSize As Integer = 13

        Private _checked As Boolean

        Private _Shown As Boolean = False

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String = ""

        Public Property Checked As Boolean
            Get
                Return _checked
            End Get
            Set(value As Boolean)
                If Not _checked.Equals(value) Then
                    _checked = value
                    Me.OnCheckedChanged()
                End If
            End Set
        End Property

        Protected Overridable Sub OnCheckedChanged()
            RaiseEvent CheckedChanged(Me, EventArgs.Empty)

            If Not DesignMode And _Shown And AnimateOnClick Then
                If Checked Then

                    Dim s As Integer = (Width - 17) * 0.5
                    For i As Integer = CheckC.Left To Width - 17 Step +5
                        CheckC.X = i + s
                        Threading.Thread.Sleep(1)
                        Refresh()
                        If i + s >= Width - 17 Then Exit For
                        s -= 1
                        If s < 0 Then s = 0
                    Next
                    CheckC.X = Width - 17

                Else

                    Dim s As Integer = 10
                    For i As Integer = CheckC.Left To 4 Step -5
                        CheckC.X = i - s
                        Threading.Thread.Sleep(1)
                        Refresh()
                        If i - s <= 4 Then Exit For
                        s -= 1
                        If s < 0 Then s = 0
                    Next
                    CheckC.X = 4

                End If

            Else
                If Checked Then
                    CheckC.X = Width - 17
                Else
                    CheckC.X = 4
                End If
            End If

            If DarkLight_Toggler Then
                CheckC.Width = DarkLight_TogglerSize
                CheckC.Height = DarkLight_TogglerSize
            End If

            Invalidate()
        End Sub
        Public Event CheckedChanged(sender As Object, e As EventArgs)

        Dim CheckedC As Rectangle

        Private AnimateOnClick As Boolean = False
        Protected Overrides Sub OnMouseClick(e As MouseEventArgs)
            AnimateOnClick = True
            Me.Checked = Not Me.Checked
            AnimateOnClick = False
            Me.Invalidate()
            MyBase.OnMouseClick(e)
        End Sub

        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Me.OnPaintBackground(e)
            Dim G As Graphics = e.Graphics
            G.SmoothingMode = SmoothingMode.AntiAlias
            DoubleBuffered = True

            If Parent Is Nothing Then Exit Sub

            BackColor = My.Style.Colors.Back

            G.Clear(GetParentColor)

            '################################################################################# Customizer
            Dim MainRect As New Rectangle(0, 0, Width - 1, Height - 1)
            Dim BorderColor As Color

            If My.Style.DarkMode Then BorderColor = BackColor.LightLight Else BorderColor = BackColor.Dark(0.5)

            Dim CheckColor As Color
            If MouseState = 0 Then CheckColor = My.Style.Colors.BaseColor Else CheckColor = BackColor.CB(If(My.Style.DarkMode, 0.3, -0.5))

            '#################################################################################
            Dim min As Integer = 4
            Dim max As Integer = Width - 17
            Dim val As Decimal = (CheckC.X) / max

            If val < 0 Then val = 0
            If val > 1 Then val = 1
            If CheckC.X <= min Then val = 0
            If CheckC.X >= max Then val = 1

            Dim lgbChecked, lgbNonChecked, lgborderChecked, lgborderNonChecked As LinearGradientBrush

            lgbChecked = New LinearGradientBrush(MainRect, Color.FromArgb(255 * val, My.Style.Colors.Border_Checked_Hover), Color.FromArgb(255 * val, My.Style.Colors.Back_Checked), LinearGradientMode.ForwardDiagonal)
            lgborderChecked = New LinearGradientBrush(MainRect, Color.FromArgb(255 * val, My.Style.Colors.Border_Checked), Color.FromArgb(255 * val, My.Style.Colors.Back_Checked), LinearGradientMode.BackwardDiagonal)
            lgbNonChecked = New LinearGradientBrush(MainRect, Color.FromArgb(255 * (1 - val), My.Style.Colors.Back_Checked), Color.FromArgb(255 * (1 - val), My.Style.Colors.Border_Checked_Hover), LinearGradientMode.BackwardDiagonal)
            lgborderNonChecked = New LinearGradientBrush(MainRect, Color.FromArgb(255 * (1 - val), My.Style.Colors.Border_Checked), Color.FromArgb(255 * (1 - val), My.Style.Colors.Back_Checked), LinearGradientMode.ForwardDiagonal)

            If Not DarkLight_Toggler Then

                Using br As New SolidBrush(Color.FromArgb(255 * val, My.Style.Colors.Border_Checked_Hover)) : e.Graphics.FillRoundedRect(br, MainRect, 9, True) : End Using

                Using P As New Pen(Color.FromArgb(255 * val, My.Style.Colors.Border_Checked)) : e.Graphics.DrawRoundedRect(P, MainRect, 9, True) : End Using

                Using P As New Pen(Color.FromArgb(255 * (1 - val), BorderColor)) : e.Graphics.DrawRoundedRect(P, MainRect, 9, True) : End Using

                If Checked Then
                    Using br As New SolidBrush(If(My.Style.DarkMode, Color.Black, Color.White)) : G.FillEllipse(br, CheckC) : End Using
                Else
                    Using br As New SolidBrush(BorderColor) : G.FillEllipse(br, CheckC) : End Using
                End If

            Else

                e.Graphics.FillRoundedRect(lgbChecked, MainRect, 9, True)
                e.Graphics.FillRoundedRect(lgbNonChecked, MainRect, 9, True)

                If Checked Then
                    G.DrawImage(If(BorderColor.IsDark, My.Resources.darkmode_dark, My.Resources.darkmode_light).Fade(val), CheckC)
                    G.DrawImage(If(BorderColor.IsDark, My.Resources.lightmode_dark, My.Resources.lightmode_light).Fade(1 - val), CheckC)
                Else
                    G.DrawImage(If(My.Style.Colors.BaseColor.IsDark, My.Resources.darkmode_dark, My.Resources.darkmode_light).Fade(val), CheckC)
                    G.DrawImage(If(My.Style.Colors.BaseColor.IsDark, My.Resources.lightmode_dark, My.Resources.lightmode_light).Fade(1 - val), CheckC)
                End If

                Using P As New Pen(lgborderChecked) : e.Graphics.DrawRoundedRect(P, MainRect, 9, True) : End Using
                Using P As New Pen(lgborderNonChecked) : e.Graphics.DrawRoundedRect(P, MainRect, 9, True) : End Using
            End If
        End Sub

        Private Sub Toggle_Resize(sender As Object, e As EventArgs) Handles Me.Resize
            Me.Height = 20
            If Width < 40 Then Width = 40

            If DarkLight_Toggler Then
                CheckC.Width = DarkLight_TogglerSize
                CheckC.Height = DarkLight_TogglerSize
            End If

            Refresh()
        End Sub

        Private Sub Toggle_HandleCreated(sender As Object, e As EventArgs) Handles Me.HandleCreated

            If Checked Then
                CheckC = New Rectangle(Width - 17, 4, 11, 11)
            Else
                CheckC = New Rectangle(4, 4, 11, 11)
            End If

            If DarkLight_Toggler Then
                CheckC.Width = DarkLight_TogglerSize
                CheckC.Height = DarkLight_TogglerSize
            End If

            If Not DesignMode Then
                Try
                    AddHandler FindForm.Load, AddressOf Loaded
                    AddHandler FindForm.Shown, AddressOf Showed
                    AddHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                Catch
                End Try
            End If
        End Sub

        Private Sub Toggle_HandleDestroyed(sender As Object, e As EventArgs) Handles Me.HandleDestroyed
            If Not DesignMode Then
                Try
                    RemoveHandler FindForm.Load, AddressOf Loaded
                    RemoveHandler FindForm.Shown, AddressOf Showed
                    RemoveHandler Parent.BackColorChanged, AddressOf RefreshColorPalette
                Catch
                End Try
            End If
        End Sub

        Sub Loaded()
            _Shown = False
        End Sub

        Sub Showed()
            _Shown = True
            Invalidate()
        End Sub

        Public Sub RefreshColorPalette()
            If _Shown Then
                Invalidate()
            End If
        End Sub

        Private Sub Toggle_MouseMove(sender As Object, e As MouseEventArgs) Handles Me.MouseMove
            If e.Button = MouseButtons.Left Then

                Dim i As Integer = e.X - 0.5 * CheckC.Width
                WasMoving = True
                MouseState = 1

                If i <= 4 Then
                    CheckC.X = 4
                ElseIf i >= Width - 17 Then
                    CheckC.X = Width - 17
                Else
                    CheckC.X = i
                End If

                CheckC.Y = 4
                Refresh()
            End If
        End Sub

        Private Sub Toggle_MouseUp(sender As Object, e As MouseEventArgs) Handles Me.MouseUp
            MouseState = 0
            CheckC.Width = 11

            If DarkLight_Toggler Then
                CheckC.Width = DarkLight_TogglerSize
                CheckC.Height = DarkLight_TogglerSize
            End If

            If WasMoving Then
                If e.X < Width * 0.5 Then Checked = False
                If e.X >= Width * 0.5 Then Checked = True
                WasMoving = False
            End If
            Refresh()
        End Sub

        Private Sub Toggle_MouseDown(sender As Object, e As MouseEventArgs) Handles Me.MouseDown
            MouseState = 1
            CheckC.Width = 13

            Refresh()
        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/Checkers/Toggle.cs
using FluentTransitions;
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using WinPaletter.Properties;

namespace WinPaletter.UI.WP
{
    [Description("Themed toggle for WinPaletter UI")]
    [DefaultEvent("CheckedChanged")]
    public class Toggle : Control
    {
        public Toggle()
        {
            SetStyle(ControlStyles.UserPaint |  ControlStyles.SupportsTransparentBackColor |  ControlStyles.OptimizedDoubleBuffer |  ControlStyles.AllPaintingInWmPaint |  ControlStyles.ResizeRedraw, true);
            DoubleBuffered = true;
            BackColor = Color.Transparent;

            Size = new(40, 20);
            Text = string.Empty;

            _checkerX = Checked ? Width - 17 : 4;
            CheckC = new(CheckerX, 4, 11, 11);

            if (DarkLight_Toggler)
            {
                CheckC.Width = DarkLight_TogglerSize;
                CheckC.Height = DarkLight_TogglerSize;
            }
        }

        #region Variables

        private bool CanAnimate => !DesignMode && Program.Style.Animations && this != null && Visible && Parent != null && Parent.Visible && FindForm() != null && FindForm().Visible;

        private RectangleF CheckC = new(4, 4, 11, 11);
        private bool WasMoving = false;
        private readonly int DarkLight_TogglerSize = 13;
        
        private Rectangle _mainRect;
        private int _min = 4;
        private int _max;
        private float _lastAlpha = -1;
        private bool _lastDarkLightMode = false;
        private bool _lastCheckedState = false;

        #endregion

        #region Properties

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public new Color BackColor { get => Color.Transparent; set {; } }

        public bool DarkLight_Toggler { get; set; } = false;

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(MultilineStringEditor), typeof(UITypeEditor))]
        [Bindable(true)]
        public override string Text { get; set; } = string.Empty;

        private bool _checked;
        public bool Checked
        {
            get => _checked;
            set
            {
                if (value != _checked)
                {
                    _checked = value;
                    OnCheckedChanged();
                }
            }
        }

        private float _checkerX = 4;
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public float CheckerX
        {
            get => _checkerX;
            set 
            { 
                if (_checkerX != value)
                {
                    _checkerX = value; 
                    Invalidate();
                }
            }
        }

        #endregion

        #region Events/Overrides

        public event EventHandler CheckedChanged;

        protected virtual void OnCheckedChanged()
        {
            CheckedChanged?.Invoke(this, EventArgs.Empty);

            if (CanAnimate)
            {
                Transition.With(this, nameof(CheckerX), _checked ? Width - 17f : 4f).CriticalDamp(Program.AnimationSpan_Quick);
            }
            else
            {
                _checkerX = Checked ? Width - 17 : 4;
                Invalidate();
            }

            if (DarkLight_Toggler)
            {
                CheckC.Width = DarkLight_TogglerSize;
                CheckC.Height = DarkLight_TogglerSize;
                Invalidate();
            }
        }

        protected override void OnResize(EventArgs e)
        {
            Height = 20;
            if (Width < 40) Width = 40;

            _mainRect = new Rectangle(0, 0, Width - 1, Height - 1);
            _max = Width - 17;

            if (DarkLight_Toggler)
            {
                CheckC.Width = DarkLight_TogglerSize;
                CheckC.Height = DarkLight_TogglerSize;
            }

            Invalidate();

            base.OnResize(e);
        }

        protected override void OnMouseMove(MouseEventArgs e)
        {
            if (e.Button == MouseButtons.Left)
            {
                float i = e.X - 0.5f * CheckC.Width;

                // Only consider it a drag if it moves more than 2 pixels
                if (i - CheckerX > 2f) WasMoving = true;

                float newCheckerX;
                if (i <= 4) newCheckerX = 4;
                else if (i >= Width - 17) newCheckerX = Width - 17;
                else newCheckerX = i;

                if (Math.Abs(_checkerX - newCheckerX) > 0.1f)
                {
                    _checkerX = newCheckerX;
                    CheckC.X = CheckerX;
                    CheckC.Y = 4;
                    Invalidate();
                }
            }

            base.OnMouseMove(e);
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            CheckC.Width = DarkLight_Toggler ? DarkLight_TogglerSize : 11;
            CheckC.Height = DarkLight_Toggler ? DarkLight_TogglerSize : 11;

            // Only consider it a drag if it actually moved more than a threshold (e.g., 2px)
            if (WasMoving)
            {
                Checked = e.X >= Width * 0.5f;
                WasMoving = false;
            }
            else
            {
                // Otherwise, toggle on simple click
                Checked = !Checked;
            }

            Invalidate();

            base.OnMouseUp(e);
        }

        protected override void OnMouseDown(MouseEventArgs e)
        {
            if (CheckC.Width != 13)
            {
                CheckC.Width = 13;
                Invalidate();
            }

            base.OnMouseDown(e);
        }

        int parentLevel = 0;
        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);

            parentLevel = this.Level();
        }

        #endregion

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            //Leave it empty to make control background transparent
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.AntiAlias;

            //Makes background drawn properly, and transparent
            InvokePaintBackground(this, e);

            // Initialize cached values if needed
            if (_mainRect.Width != Width - 1 || _mainRect.Height != Height - 1)
            {
                _mainRect = new Rectangle(0, 0, Width - 1, Height - 1);
                _max = Width - 17;
            }

            Config.Scheme scheme = Enabled ? Program.Style.Schemes.Main : Program.Style.Schemes.Disabled;

            CheckC.X = CheckerX;

            // Calculate alpha value - optimized to avoid redundant calculations
            float val = (float)(CheckC.X / _max);
            if (val < 0f) val = 0f;
            if (val > 1f) val = 1f;
            if (CheckC.X <= _min) val = 0f;
            if (CheckC.X >= _max) val = 1f;

            int alpha = (int)(255f * val);

            // Skip painting if nothing changed (except for animations)
            if (!CanAnimate && alpha == _lastAlpha && DarkLight_Toggler == _lastDarkLightMode && Checked == _lastCheckedState) return;

            _lastAlpha = alpha;
            _lastDarkLightMode = DarkLight_Toggler;
            _lastCheckedState = Checked;

            if (!DarkLight_Toggler)
            {
                // Checked part - reuse brushes to reduce allocations
                Color accentColor = Color.FromArgb(alpha, scheme.Colors.AccentAlt);
                Color foreColor = Color.FromArgb(255 - alpha, scheme.Colors.ForeColor);
                Color checkerColor = Color.FromArgb(alpha, scheme.Colors.AccentAlt.IsDark() ? Color.Black : Color.White);

                using (SolidBrush accentBrush = new(accentColor))
                using (SolidBrush foreBrush = new(foreColor))
                using (SolidBrush checkerBrush = new(checkerColor))
                using (Pen accentPen = new(accentColor))
                using (Pen forePen = new(foreColor))
                {
                    G.FillRoundedRect(accentBrush, _mainRect, 9, true);
                    G.DrawRoundedRectBeveled(accentPen, _mainRect, 9, true);
                    G.FillEllipse(checkerBrush, CheckC);
                    G.FillEllipse(foreBrush, CheckC);
                    G.DrawRoundedRect(forePen, _mainRect, 9, true);
                }
            }
            else
            {
                // DarkLight mode - optimize LinearGradientBrush creation
                Color checkedColor1 = Color.FromArgb(alpha, scheme.Colors.Accent);
                Color checkedColor2 = Color.FromArgb(alpha, scheme.Colors.AccentAlt);
                Color nonCheckedColor1 = Color.FromArgb(255 - alpha, scheme.Colors.AccentAlt);
                Color nonCheckedColor2 = Color.FromArgb(255 - alpha, scheme.Colors.Accent);

                using (LinearGradientBrush lgbChecked = new(_mainRect, checkedColor1, checkedColor2, LinearGradientMode.ForwardDiagonal))
                using (LinearGradientBrush lgborderChecked = new(_mainRect, checkedColor1, checkedColor2, LinearGradientMode.BackwardDiagonal))
                using (LinearGradientBrush lgbNonChecked = new(_mainRect, nonCheckedColor1, nonCheckedColor2, LinearGradientMode.BackwardDiagonal))
                using (LinearGradientBrush lgborderNonChecked = new(_mainRect, nonCheckedColor1, nonCheckedColor2, LinearGradientMode.ForwardDiagonal))
                {
                    G.FillRoundedRect(lgbChecked, _mainRect, 9, true);
                    G.FillRoundedRect(lgbNonChecked, _mainRect, 9, true);

                    // Optimize bitmap operations
                    Bitmap baseBitmap = scheme.Colors.Line(parentLevel).IsDark() ? Resources.darkmode_light : Resources.darkmode_dark;
                    Bitmap overlayBitmap = scheme.Colors.Line(parentLevel).IsDark() ? Resources.lightmode_light : Resources.lightmode_dark;

                    if (!Checked)
                    {
                        baseBitmap = scheme.Colors.AccentAlt.IsDark() ? Resources.darkmode_light : Resources.darkmode_dark;
                        overlayBitmap = scheme.Colors.AccentAlt.IsDark() ? Resources.lightmode_light : Resources.lightmode_dark;
                    }

                    using (Bitmap b0 = baseBitmap.Fade(val))
                    using (Bitmap b1 = overlayBitmap.Fade(1f - val))
                    {
                        G.DrawImage(b0, CheckC);
                        G.DrawImage(b1, CheckC);
                    }

                    using (Pen P = new(lgborderChecked)) { G.DrawRoundedRectBeveled(P, _mainRect, 9, true); }
                    using (Pen P = new(lgborderNonChecked)) { G.DrawRoundedRectBeveled(P, _mainRect, 9, true); }
                }
            }

            base.OnPaint(e);
        }
    }
}